lib/src/source/hosted.dart

```diff
@@ -144,35 +145,68 @@ class BoundHostedSource extends CachedSource {
   final HostedSource source;
 
   final SystemCache systemCache;
+  Retriever<PackageRef, Map<PackageId, Pubspec>> _retriever;
 
-  BoundHostedSource(this.source, this.systemCache);
+  BoundHostedSource(this.source, this.systemCache) {
+    _retriever =
+        Retriever.nonCancelable(_getVersions, maxConcurrentOperations: 10);
+  }
 
-  /// Downloads a list of all versions of a package that are available from the
-  /// site.
-  Future<List<PackageId>> doGetVersions(PackageRef ref) async {
+  Future<Map<PackageId, Pubspec>> _getVersions(
+      PackageRef ref, Retriever retriever) async {
     var url = _makeUrl(
         ref.description, (server, package) => "$server/api/packages/$package");
-
     log.io("Get versions from $url.");
-
     String body;
     try {
+      // TODO(sigurdm): Implement cancellation of requests. This probably
+      // requires resolution of: https://github.com/dart-lang/sdk/issues/22265.
       body = await httpClient.read(url, headers: pubApiHeaders);
     } catch (error, stackTrace) {
       var parsed = source._parseDescription(ref.description);
       _throwFriendlyError(error, stackTrace, parsed.first, parsed.last);
     }
-
-    var doc = jsonDecode(body);
-    return (doc['versions'] as List).map((map) {
+    final doc = jsonDecode(body);
+    final result = Map.fromEntries((doc['versions'] as List).map((map) {
       var pubspec = Pubspec.fromMap(map['pubspec'], systemCache.sources,
           expectedName: ref.name, location: url);
       var id = source.idFor(ref.name, pubspec.version,
           url: _serverFor(ref.description));
-      memoizePubspec(id, pubspec);
+      return MapEntry(id, pubspec);
+    }));
+
+    // Prefetch the dependencies of the latest version, we are likely to need
+    // them later.
+    void prefetch() {
+      final latest = doc['latest'];
```

Comment:
Should we consider using `result.values` which probably can be sorted based on version numbers, as `pub_semver` knows how to compare these..

This avoids us depending on a property from the server that we didn't depend on before. It also ensure that we use the same ordering as the solve uses...

This is a small thing, and maybe I'm just crazy, just thinking that we should consider if we can avoid trusting the server for anything new :)

Reply:
I think you are right that might be the more safe thing to do.
Done